### PR TITLE
Read-only instance serial console API endpoint in Nexus.

### PR DESCRIPTION
**Pull request by lifning**
This first-pass implementation is simply a GET with parameters for
byte offset since boot and length, with the intent of providing something
simple for the web frontend to use. It's implemented by cacheing both the
first MiB and most recent MiB of serial console output in the sled-agent,
though this detail is not important in itself (i.e. it could just as well
have the meat of the implementation moved to propolis-server.)

NB: Requires https://github.com/oxidecomputer/propolis/pull/131 (so won't pass checks without it)

### CODE REVIEW

**Comment by david-crespo** [comment]
Very excited about this. What's an example of a situation where someone would want to use a negative offset?

**Comment by lifning** [comment]
> Very excited about this. What's an example of a situation where someone would want to use a negative offset?

Supposing the live view of the console wants to fetch the most recent N bytes when it loads, and my guess of 16KiB as a default for that ends up being over/under-kill for "request too big" or "doesn't go back far enough for XYZ usecase" reasons.

File: common/src/api/internal/nexus.rs
Comment (david-crespo) [inline common/src/api/internal/nexus.rs:111]: Is the "since instance boot" here accurate? I can't square that with negative offsets.

File: common/src/api/internal/nexus.rs
Comment (david-crespo) [inline common/src/api/internal/nexus.rs:121]: `InstanceSerialConsoleRequest` ends up being used by both Sled Agent's internal API and Nexus's external API, so `common/api/internal/nexus.rs` is an uncomfortable spot for it. Since it's so simple and doesn't depend on any other structs (and there's no essential requirement that it be identical between Nexus and Sled Agent) I lean toward giving Nexus and Sled Agent each their own copy in these two files:

https://github.com/oxidecomputer/omicron/blob/main/nexus/src/external_api/params.rs
https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/params.rs

`InstanceSerialConsoleData` is used in more spots and it's a bit harder for me to picture how it should look, but it probably could use the same treatment. Then when, for example, Nexus needs to refer to Sled Agent's response type, it can refer to the version on the generated Sled Agent client rather than the actual original struct used by Sled Agent.

File: sled-agent/src/serial.rs
Comment (luqmana) [inline sled-agent/src/serial.rs:178]: I don't remember if the tungstenite crate handles it transparently, but would this also try to "consume" the WebSocket ping/pong messages?

File: sled-agent/src/serial.rs
Comment (luqmana) [inline sled-agent/src/serial.rs:191]: Up to you if you want to duplicate the comment or not, but it's nice to have the "entrypoints" to a mod/type be documented.

File: sled-agent/src/sim/sled_agent.rs
Comment (luqmana) [inline sled-agent/src/sim/sled_agent.rs:206]: The legend of localhorse continues @jordanhendricks  😅️

File: sled-agent/src/instance.rs
Comment (luqmana) [inline sled-agent/src/instance.rs:613]: I'd move this into `InstanceInner::ensure` to better match the current instance setup

File: nexus/src/app/instance.rs
Comment (luqmana) [inline nexus/src/app/instance.rs:1044]: I'm assuming this is because we use `usize`/`isize` in the rust types but the generated OpenAPI stuff maps that to `u32`/`i32`? If so, would it be clearer to just use the explicitly sized ones as well in the types?

**Review by luqmana — CHANGES_REQUESTED** [review]

File: sled-agent/src/serial.rs
Comment (lifning) [reply sled-agent/src/serial.rs:32]: took another look at it and decided to simply localize the `isize`-ness to the one function where it's relevant.

File: sled-agent/src/serial.rs
Comment (lifning) [reply sled-agent/src/serial.rs:178]: totally would, oops! it *does* handle sending pongs transparently, but i think they're also passed through.

File: nexus/src/app/instance.rs
Comment (lifning) [reply nexus/src/app/instance.rs:1044]: yeah, this was the rustaceanary urge to use usize/isize for array length / indexing, but if that's the right thing to do in the OpenAPI world i can certainly change that and do the conversions in the other direction where needed.

File: sled-agent/src/sim/sled_agent.rs
Comment (lifning) [reply sled-agent/src/sim/sled_agent.rs:206]: (OS/478 was kind of my take on a fictionalized far-future OS/2 successor, the `0x1DE`-th in the franchise)

File: sled-agent/src/serial.rs
Comment (smklein) [inline sled-agent/src/serial.rs:1]: Nit: Let's add the MPL header

File: common/src/api/internal/nexus.rs
Comment (smklein) [inline common/src/api/internal/nexus.rs:113]: I'm trying to wrap my head around the different uses of this request structure, and I think I'm struggling :sweat_smile: 

- What would `None` mean here?
- In the positive vs negative cases, we are using two entirely different frames-of-reference, right? Like, "0" always means "the first byte emitted from the instance", but "-1" means "the most recent byte we've seen" (which *could* be equivalent to a positive value `byte_offset`, depending on what has been buffered)? I'm a little surprised by this non-linear indexing usage.

Does an `Option<isize>` actually make sense for this dual-reference indexing? WDYT about something like this:

```rust
enum Offset {
  // Index from the start of the instance.
  FromStart(usize),

  // Return the most recent bytes seen from the instance's serial console.
  MostRecent(usize),
}
```

**Review by smklein — COMMENTED** [review]
This integration looks good to me! Nice work on such a cross-stack feature.

Before giving the stamp of approval, I have a couple questions related to indexing, and "what requests will results in errors vs truncated buffers". More detail below.

File: common/src/api/internal/nexus.rs
Comment (lifning) [reply common/src/api/internal/nexus.rs:113]: I'm definitely in favor of doing this in the underlying code, I was kind of inclined to avoid it at the API surface itself for https://github.com/oxidecomputer/omicron/issues/117 reasons

File: common/src/api/internal/nexus.rs
Comment (lifning) [reply common/src/api/internal/nexus.rs:126]: it would return an error if we are requesting a range that *begins* somewhere we don't have cached.

for positive values of `byte_offset`, the returned `last_byte_offset` is effectively `byte_offset + data.len()`, the intent is that the client gets an easy value to plug into the next call's `byte_offset`. (this is more useful when indexing from most-recent, but still potentially convenient when indexing from beginning)

File: tools/serialconsole_demo.sh
Comment (lifning) [reply tools/serialconsole_demo.sh:11]: i'll write the more-generally-usable version into the oxide CLI

File: common/src/api/internal/nexus.rs
Comment (zephraph) [reply common/src/api/internal/nexus.rs:113]: Enums like this aren't really a problem. The problem mostly arises in enums that mix string types with richer data structures. This is less of an issue today now that we have our own generator than it was initially though, so I think it's fine either way. cc @david-crespo to check me on that. 

In general designing the api to be more expressive and explicit is probably better from a usability perspective.

Comment (david-crespo) [reply common/src/api/internal/nexus.rs:113]: Yep, that's correct. I think the enum would turn out fine. I believe in the TS client it would turn into something very usable like this:

```ts
type Offset = 
 | { type: "FromStart"; value: number }
 | { type: "MostRecent"; value: number }
```

This is partly because our generator is better than the the off the shelf options for our purposes, but I also think Dropshot's OpenAPI output has been improved.

Comment (lifning) [reply common/src/api/internal/nexus.rs:113]: So I've been running into https://github.com/oxidecomputer/dropshot/blob/784399af2473906f96e2c49698c3fd94933c57b7/dropshot/src/type_util.rs#L71 when I put the tagged enum straight into the endpoint - am I missing something obvious?

Comment (david-crespo) [reply common/src/api/internal/nexus.rs:113]: Not sure without seeing how you're doing it, but here are a couple that work:

https://github.com/oxidecomputer/omicron/blob/e2bd1e85ccad60ff9b58c751929fa2c51ce9b6b7/nexus/src/external_api/params.rs#L331-L358

https://github.com/oxidecomputer/omicron/blob/e2bd1e85ccad60ff9b58c751929fa2c51ce9b6b7/sled-agent/src/params.rs#L28-L36

Comment (smklein) [reply common/src/api/internal/nexus.rs:113]: @lifning - if you have a branch that includes the enum (even if it makes dropshot unhappy), I'd be happy to help debug?

Comment (lifning) [reply common/src/api/internal/nexus.rs:113]: https://github.com/lifning/omicron/tree/wip is where my branch is, if you wanna take a look

Comment (smklein) [reply common/src/api/internal/nexus.rs:113]: Hey, so I checked that branch out - I'm seeing errors while building, because there's a dependency from `omicron-sled-agent` --> `propolis-client`, which appears to be pointing to a revision of propolis that does not exist.

See: https://github.com/lifning/omicron/blame/wip/sled-agent/Cargo.toml#L27

This change looks like it comes from https://github.com/lifning/omicron/commit/0936a92aa4fb9b3cc133415e2f1ca836ac8b1825#diff-46cfcab5a82aa792a22470d4b5f050a5d7c9618ec8c44971ab7cde19911fe4ff 

Would it be a problem to just use the latest revision of Propolis here?

Comment (smklein) [reply common/src/api/internal/nexus.rs:113]: Ah, I think this may be the dependency on https://github.com/oxidecomputer/propolis/pull/131

Comment (lifning) [reply common/src/api/internal/nexus.rs:113]: Yeah--is it okay to merge that ahead of this being ready to go in to actually use it?

Comment (luqmana) [reply common/src/api/internal/nexus.rs:113]: Yea, I'd say go for it. It's small and self-contained. You can still make changes after that as needed.

Comment (david-crespo) [reply common/src/api/internal/nexus.rs:113]: Now that I understand [the issue](https://github.com/oxidecomputer/dropshot/issues/369) better, I think the problem is that this about a query string, and Dropshot doesn't know how you would represent this enum as a query string. When I agreed the enum was fine, I thought we were talking about a POST request body or a response body. Sorry!

In order to make this work in the query string it would need to look more like a simple struct (like it was before haha):

```rs
struct QueryArgs {
  from_start_offset: Option<usize>,
  from_end_offset: Option<usize>
}
```

so it can get turned into `?from_start_offset=100` etc.

Comment (lifning) [reply common/src/api/internal/nexus.rs]: luckily enough, that's the first thing i did before spending a couple days banging my head on getting it to accept the enum it -- so i simply dropped that commit from my branch :)

File: nexus/src/external_api/params.rs
Comment (david-crespo) [reply nexus/src/external_api/params.rs:206]: Or I might even leave it out. The similarity is sort of incidental and pretty well expressed at the contact point between Nexus and Sled Agent:

https://github.com/oxidecomputer/omicron/blob/37555544820041200e1751efc1f9a9a83021c91f/nexus/src/app/instance.rs#L850-L864

File: sled-agent/src/sim/sled_agent.rs
Comment (david-crespo) [inline sled-agent/src/sim/sled_agent.rs:170]: 😁

File: test-utils/src/dev/test_cmds.rs
Comment (luqmana) [inline test-utils/src/dev/test_cmds.rs:43]: Nit: I would just use `&str` as the type here. I don't think we need this level of genericism here since all the call sites seem to just be doing `&foo` where `foo` is a `String`.

Comment (luqmana) [inline test-utils/src/dev/test_cmds.rs:43]: Nit: I would just use `&str` as the type here. I don't think we need this level of genericism here since all the call sites seem to just be doing `&foo` where `foo` is a `String`.

File: nexus/src/external_api/params.rs
Comment (luqmana) [inline nexus/src/external_api/params.rs:422]: 💯 on the great documentation.

File: sled-agent/src/serial.rs
Comment (luqmana) [inline sled-agent/src/serial.rs:160]: Nit: thread->task

**Review by luqmana — APPROVED** [review]
Woo! The instance stuff looks good to me. I'll defer to others for any wider omicron comments.

Like Sean said, this cuts across different parts of the stack. Thanks for driving this through lif!

**Comment by lifning** [comment]
I guess one question I have is - should I rework the semantics of how the numerical parameters are handled to be more consistent with the opinionation of the oxide.rs bindings generator? (i.e. no such thing as an Option<u64>, 0 means None)

**Pull request merged**
